---
title: "________________________________________________"
subtitle: "Publications using CeMEE populations and RILs"
output: html_document
bibliography: /Users/noble/Documents/github/cemee/assets/cemee2.bib
---



* [Teotonio et al.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0035811) [-@Teotonio2012-xj] 
Derivation of the ancestral **A0** population and adaptation of replicate populations varying in mating system to the lab environment for 100 generations. 

* [Chelo et al.](https://www.nature.com/articles/ncomms3417?origin=ppub) [-@Chelo2013-lm] 
Testing of frequency-dependent allelic extinction probabilities using two inbred lines derived from **A6140**, one after introgression of an integrated GFP reporter under the *myo-3* promoter.

* [Chelo & Teotonio](https://onlinelibrary.wiley.com/doi/full/10.1111/j.1558-5646.2012.01744.x) [-@Chelo2013-lp] Individual-level genotyping from dioecious and androdioecious populations derived from **A0** to study the mode of selection underlying maintenance of genetic diversity during adaptation.

* [Goussen et al.](https://doi.org/10.1007/s10646-013-1078-5) [-@Goussen2013-kf] 
The veolution of resistance to Uranium, using **A0**.

* [Chelo et al.](https://www.nature.com/articles/hdy2013100) [-@Chelo2013-ne] 
The evolution of inbreeding and outbreeding depression under androdioecy and dioecy, using populations derived from **A0**.

* [Masri et al.](https://onlinelibrary.wiley.com/doi/full/10.1111/ele.12068) [-@Masri2013-wo]
The role, and differential sensitivity, of males in evolution of resistance to *Bacillus thuringiensis*, using **A0**.

* [Carvalho et al.](https://bmcevolbiol.biomedcentral.com/articles/10.1186/1471-2148-14-116) [-@Carvalho2014-ps]
Evolution of population fitness and hermaphrodite fitness components in inbred populations derived from **A0** and adapted to the lab for 100 generations. 

* [Carvalho et al.](https://bmcevolbiol.biomedcentral.com/articles/10.1186/1471-2148-14-117) [-@Carvalho2014-tj]
Evolution of sex allocation in hermaphrodites, using replicates of dioecious and androdioecious populations derived from **A0** and adapted to the lab for 100 generations. 

* [Theologidis et al.](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-014-0093-1) [-@Theologidis2014-oc] 
Testing of hermaphrodite reproductive assurance during adaptation to a novel environment. Mono-, di- and trioecious populations were derived from **A6140** and adapted to lab conditions, or to gradual or sudden increases in NaCl (including **CA[1-3]50**, **GA[1,2,4]50**, **GT[1,2]50**, **GM[1,3]50**).

* [Masri et al.](http://dx.doi.org/10.1371/journal.pbio.1002169) [-@Masri2015-qf]
Experimental co-evolution of *Bacillus thuringiensis* and **A0**.

* [Poullet et al.](http://onlinelibrary.wiley.com/doi/abs/10.1111/evo.13032) [-@Poullet2016-vx] 
Life-history responses to experimental evolution in hermaphrodites, using **A0** and derived androdioecious populations (A[4-6]100). 

* [Dey et al.](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1002388) [-@Dey2016-ak] Experimental evolution of maternal effects, using a salt-adapted population **GA250**.

* [Parrish et al.](https://doi.org/10.1371/journal.pone.0154463) [-@Parrish2016-rr]
Experimental evolution using **A0** testing the effects of outcrossing and genetic variation on adaptation to the parasite *Serratia marcescens*.

* [Noble et al.](http://www.genetics.org/content/207/4/1663) [-@Noble2017-gh] 
Description of the CeMEE v1 RIL panel (RILs sampled from **A6140** and derived populations), and QTL mapping of fertility and worm area on NGM. 

* [Guzella et al.](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007731) [-@Guzella2018-hb] 
Evolution under gradual and rapid environmental change, using monoecious populations derived from **A6140** adapting to salt (including **GM[1-7]50**).

* [Chelo et al.](https://www.g3journal.org/content/9/9/2811) [-@Chelo2019-ki] 
The effects of mating system on genetic load, using populations derived from **A6140** and adapted to salt (including **GM[1,3]50**, **GA[1,2,4]50**, **GT[1,2]50**).

* [Noble et al.](https://www.biorxiv.org/content/10.1101/589432v1) [-@Noble2019-kf] 
Expansion of the CeMEE RIL panel, with new populations (**CA[1-3]100**), updated SNV genotypes, haplotype reconstructions, *de novo* mutation, and candidate loci under selection. 

* [Papkou et al.](http://dx.doi.org/10.1073/pnas.1810402116) [-@Papkou2019-du]
Coevolution of *Bacillus thuringiensis* and CeMEE populations.

* [Mallard, Noble, et al.](https://www.biorxiv.org/content/10.1101/778282v1.full) [-@Mallard2019-qd] 
Analysis of phenotypic evolution in populations under a constant environment (**CA[1-3]50** and **CA[1-3]100**, derived from **A6140**).

* [Archer et al.](https://www.biorxiv.org/content/10.1101/863878v1.full) [-@Archer2019-rm]
Application of the Starvation Arena to study adult starvation resistance in 72 CeMEE RILs.

* [Noble et al.](https://www.g3journal.org/content/10/7/2385.full) [-@Noble2020-st] Interaction between a *col-182* mutation in N2 and standing genetic variation in CeMEE RILs modifies worm locomotion and shape.

* [Pelizzola et al.](https://www.biorxiv.org/content/10.1101/2020.07.09.191924v2.full) [-@Pelizzola2020-or]
Development of a method ([R:haploSep](https://github.com/MartaPelizzola/haploSep)) for inferring haplotypes and their frequencies from time-series allele frequency data.


***

