---
title: "RIL metadata"
output: html_document
---

```{r include=F}
library(dplyr)
library(DT)
library(readxl)
library(kableExtra)
library(stringr)
library(data.table)
options(knitr.table.format = "html", knitr.kable.NA = '') 
```

```{r, echo=F}
meta <- read_excel('~/Dropbox/CeMEE_v2_meta.xlsx', sheet = 'rilMeta_verbose')
cryocapcols <- unique(meta$cap)
meta$cap <- factor(meta$cap, labels = c('green', 'blue', 'grey', 'orange', 'yellow', 'red'))
meta <- mutate_at(meta, c('plate', 'row', 'col', 'line', 'pop', 'pop rep', 'G+A0', 'G+A140', 'G inbred', 'line','v1', 'v2', 'seq loc', 'seq platform', 'seq year'), factor)
meta$`seq depth` <- round(meta$`seq depth`, 2)
meta$`perc ref/alt` <- round(meta$`perc ref/alt`, 1)
meta$`prop alt` <- round(meta$`prop alt`, 2)
# crop 'similar' column to max 3 lines
meta$similar <- sapply(meta$similar, function(x) {
  spl = tstrsplit(x, ',')
  nl = length(spl)
  if(nl>2){
    sprintf('%s and %s other%s', paste(unlist(spl)[1:2], collapse=','), nl-2, ifelse(nl>3, 's', ''))
  } else {
    x
  }
})

qc <- read_excel('~/Dropbox/CeMEE_v2_meta.xlsx', sheet = 'qc', range = "N1:O1000")
# QC column
maxhetfreq = 0.2
maxhetprop = 0.1
mincov = 0.1
addtag <- function(tag, boolv){
  o = matrix('PASS', nrow(meta))
  o[boolv] = tag
  o
}
qca = addtag('DUP', meta$line %in% qc$todrop.nobg[qc$case != 'haplotypeLL'])
qca <- cbind(qca, addtag('LOW', meta$`seq depth` < mincov))
qca <- cbind(qca, addtag('HAP', meta$line %in% qc$todrop.nobg[qc$case == 'haplotypeLL']))
qca <- cbind(qca, addtag('HET', meta$`perc ref/alt` > (maxhetfreq*100) & meta$`prop alt` > maxhetprop))
meta$`v2 QC` <- ''
meta$`v2 QC`[meta$v2==1] <- 'PASS'
dropix = apply(qca, 1, function(x) sum(x != 'PASS')>0)
meta$`v2 QC`[dropix] = unlist(apply(qca[dropix,], 1, function(x) paste0(unique(x[grep('PASS', x, invert = T)]), collapse=',')))
meta <- meta[,c(1:20,26,21:25)]

sketch = htmltools::withTags(table(
  class = "compact cell-border hover order-column stripe",
  thead(
    tr(
      th('Plate', title = 'Cryo plate number'),
      th('Row', title = 'Cryo plate row'),
      th('Col', title = 'Cryo plate col'),
      th('Cap', title = 'Cryo vial color'),
      th('Line', title = 'Unique identifier'),
      th('Pop', title = 'Population'),
      th('Pop rep', title = 'Population replicate'),
      th('line #', title = 'Line number'),
      th('G+A0', title = 'Generations of evolution from initial A0 population (excluding inbreeding)'),
      th('G+A140', title = 'Generations of evolution from A6140 population (excluding inbreeding)'),
      th('G inbred', title = 'Generations of RIL inbreeding by selfing'),
      th('Seq depth', title = 'Sequenced depth (estimated from reads mapped to the center of chromosome I with minimum MQ40)'),
      th('Ref/alt', title = 'Percentage of diallelic SNPs covered by at least three reads at which both the reference and alternate alleles are seen at least once in the RIL sequencing data'),
      th('Minor prop.', title = 'Mean proportion of reads showing the minor allele at sites where at least three reads are present and both alleles are seen'),
      th('Seq year', title = 'RIL year of genome sequencing'),
      th('Seq platform', title = 'RIL sequencing platform (Illumina or BGI)'),
      th('Seq loc.', title = 'RIL sequencing location (NYU or BGI'),
      th('Read len', title = 'PE read length (dominant sample where >1)'),
      th('v1', title = 'line in CeMEE version 1'),
      th('v2', title = 'line in CeMEE version 2'),
      th('v2 QC', title = sprintf('Quality control status: PASS, DUP (another closely related line has been retained), LOW (sequencing coverage <%sx), HET (ref & alt seen at >%s%% of sites with >= 3 reads and mean minor allele proportion > %s)', mincov, maxhetfreq*100, maxhetprop)),
      th('Fert. NGM', title = 'Phenotyped for fertility (number of viable L1s) on NGM. Value is the number of observations (adult hermaphrodites) post-QC.'),
      th('Fert. NaCl', title = 'Phenotyped for fertility (number of viable L1s) on NGM + 305mM NaCl. Value is the number of observations (adult hermaphrodites) post-QC.'),
      th('Loco. NGM', title = 'Locomotion on NGM. Value is the number of observations (tracks) post-QC.'),
      th('Loco. NaCl', title = 'Locomotion on NaCl. Value is the number of observations (tracks) post-QC.'),
      th('Similar', title = 'List of lines with >= 90% identity at diallelic markers. Only one of this group is retained in CeMEE v2, based on the amount of phenotyping and then sequence data')
    )
  )
))

datatable(meta, 
          container=sketch,
          rownames = F,
          options = list(pageLength = 25, 
                         autoWidth = T, 
                         scrollX=F,
                         initComplete = JS(
                           "function(settings, json) {",
                           "$(this.api().table().header()).css({'background-color': '#000', 'color': '#fff'});",
                           "}"
                         ),
                         columnDefs = list(
                           list(className = 'dt-center', targets='_all'),
                           list(width = '200px', targets='_all')
                         )
          ), 
          filter='top'
) %>% formatStyle('prop alt', background = styleColorBar(meta$`prop alt`, 'firebrick')) %>% formatStyle('similar', fontSize=1) %>%
  formatStyle('cap', fontSize=0, backgroundColor = styleEqual(levels = unique(meta$cap), values = paste0('#', cryocapcols))) %>%
  formatStyle('seq depth', background = styleColorBar(meta$`seq depth`, 'steelblue')) %>%
  formatStyle('perc ref/alt', background = styleColorBar(meta$`perc ref/alt`, 'firebrick'))
```