---
title: "RIL metadata"
output: html_document
---

```{r include=F}
library(dplyr)
library(DT)
library(readxl)
library(kableExtra)
library(stringr)
options(knitr.table.format = "html", knitr.kable.NA = '') 
```

Basic information for each RIL (and a few populations) is below. Note, not all RILs have been sequenced yet, and even fewer have been phenotyped. There's also a more verbose version. 

```{r, echo=F}
meta <- read_excel('~/Documents/cemee/CeMEE_meta.xltm', sheet = 'rilMeta_verbose')

sketch = htmltools::withTags(table(
  class = "compact cell-border hover order-column stripe",
  thead(
    tr(
      th('', title = 'Row'),
      th('plate', title = 'Cryo plate number'),
      th('row', title = 'Cryo plate row'),
      th('col', title = 'Cryo plate col'),
      th('cap', title = 'Cryo vial color'),
      th('line', title = 'Unique identifier'),
      th('pop', title = 'Population'),
      th('pop rep', title = 'Population replicate'),
      th('line #', title = 'Line number'),
      th('G+A140', title = 'Generations of evolution from A6140 population'),
      th('G+A0', title = 'Generations of evolution from initial A0 population'),
      th('seq depth', title = 'Sequenced depth (estimated from reads mapped to the center of chromosome I with minimum MQ40)'),
      th('perc ref/alt', title = 'Percentage of diallelic SNPs at which both the reference and alternate allele are seen in the RIL sequencing data'),
      th('minor prop', title = 'Mean proportion of reads showing the minor allele at sites where at least three reads are present and both alleles are seen'),
      th('G inbred', title = 'Generations of RIL inbreeding by selfing'),
      th('seq year', title = 'RIL year of genome sequencing'),
      th('seq platform', title = 'RIL sequencing platform (Illumina or BGI)'),
      th('seq loc', title = 'RIL sequencing location (NYU or BGI'),
      th('read len', title = 'PE read length (dominant sample where >1)'),
      th('QC', title = 'Quality control status: PASS, DUP, LOW COV'),
      th('similar lines', title = 'List of lines with >= 90% identity at diallelic markers. Only one of this group is retained in CeMEE v2, based on the amount of phenotyping and then sequence data'),
      th('pheno', title = 'List of phenotypes measured')
    )
  )
))

datatable(meta, container=sketch,
          options = list(pageLength = 50, autoWidth = TRUE,
                         initComplete = JS(
                           "function(settings, json) {",
                           "$(this.api().table().header()).css({'background-color': '#000', 'color': '#fff'});",
                           "}"),
                         columnDefs = list(list(className = 'dt-center', targets=1:21))
                         ), 
          filter='top'
) %>% formatStyle('cap', backgroundColor = styleEqual(levels = unique(meta$cap), values = paste0('#', unique(meta$cap))))
```