---
title: "RIL metadata"
output: html_document
---

```{r include=F}
library(dplyr)
library(DT)
library(readxl)
library(kableExtra)
library(stringr)
library(data.table)
options(knitr.table.format = "html", knitr.kable.NA = '') 
```

Basic information for CeMEE founders, extant cryopreserved RILs, and a few intermediate populations. Not all RILs or populations have been sequenced yet, and fewer have been phenotyped.

There's also [a more verbose version](rilMeta_verbose.html) with additional sequencing and QC metadata.

Hover over the columns for more information.


```{r, echo=F}
meta <- read_excel('~/Dropbox/CeMEE_v2_meta.xlsx', sheet = 'rilMeta_verbose')
cryocapcols <- unique(meta$cap)
meta$cap <- factor(meta$cap, labels = c('green', 'blue', 'grey', 'orange', 'yellow', 'red'))
meta <- mutate_at(meta, c('plate', 'row', 'col', 'line', 'pop', 'pop rep', 'G+A0', 'G+A140', 'G inbred', 'line','v1', 'v2', 'v2 QC', 'seq loc', 'seq platform', 'seq year'), factor)
meta$`seq depth` <- round(meta$`seq depth`, 1)

# QC column
maxhetfreq = 0.2
maxhetprop = 0.1
mincov = 0.1

meta <- meta[,c(5,21,6:8,12,22:25)]

sketch = htmltools::withTags(table(
  class = "compact cell-border hover order-column stripe",
  thead(
    tr(
      th('line', title = 'Unique identifier'),
      th('v2 QC', title = sprintf('Quality control status: PASS, DUP (another closely related line has been retained), LOW COV (<%sx), HET (ref & alt seen at >%s%% of sites with >= 3 reads and mean minor allele proportion > %s)', mincov, maxhetfreq*100, maxhetprop)),
      th('pop', title = 'Population'),
      th('pop rep', title = 'Population replicate'),
      th('line #', title = 'Line number'),
      th('seq depth', title = 'Sequenced depth (estimated from reads mapped to the center of chromosome I with minimum MQ40)'),
      th('fert. NGM', title = 'Phenotyped for fertility (number of viable L1s) on NGM. Value is the number of observations (adult hermaphrodites) post-QC.'),
      th('fert. NaCl', title = 'Fertility (number of viable L1s) on 300mM NaCl. Value is the number of observations (adult hermaphrodites) post-QC.'),
      th('loc. NGM', title = 'Locomotion on NGM. Value is the number of observations (tracks) post-QC.'),
      th('loc. NaCl', title = 'Locomotion on NaCl. Value is the number of observations (tracks) post-QC.')
    )
  )
))

datatable(meta, 
          container=sketch,
          rownames = F,
          options = list(pageLength = 25, 
                         autoWidth = T, 
                         scrollX=F,
                         initComplete = JS(
                           "function(settings, json) {",
                           "$(this.api().table().header()).css({'background-color': '#000', 'color': '#fff'});",
                           "}"
                         ),
                         columnDefs = list(
                           list(className = 'dt-center', targets='_all'),
                           list(width = '200px', targets='_all')
                         )
          ), 
          filter='top'
) %>% formatStyle('seq depth', background = styleColorBar(meta$`seq depth`, 'steelblue'))
```