---
title: "RIL metadata"
output: html_document
---

```{r include=F}
library(dplyr)
library(DT)
library(readxl)
library(kableExtra)
library(stringr)
library(data.table)
options(knitr.table.format = "html", knitr.kable.NA = '') 
```

Basic information for CeMEE founders, extant cryopreserved RILs, and a few intermediate populations. Not all RILs or populations have been sequenced yet, and fewer have been phenotyped.

There's also [a more verbose version](rilMeta_verbose.html) with additional sequencing, QC and cryo metadata.

Hover over the columns for more information.


```{r, echo=F}
meta <- read_excel('~/Dropbox/CeMEE_v2_meta.xlsx', sheet = 'rilMeta_verbose')
cryocapcols <- unique(meta$cap)
meta$cap <- factor(meta$cap, labels = c('green', 'blue', 'grey', 'orange', 'yellow', 'red'))
meta <- mutate_at(meta, c('plate', 'row', 'col', 'line', 'pop', 'pop rep', 'G+A0', 'G+A140', 'G inbred', 'line','v1', 'v2', 'seq loc', 'seq platform', 'seq year'), factor)
meta$`seq depth` <- round(meta$`seq depth`, 2)
meta$`perc ref/alt` <- round(meta$`perc ref/alt`, 1)
meta$`prop alt` <- round(meta$`prop alt`, 2)
# crop 'similar' column to max 3 lines
meta$similar <- sapply(meta$similar, function(x) {
  spl = tstrsplit(x, ',')
  nl = length(spl)
  if(nl>2){
    sprintf('%s and %s other%s', paste(unlist(spl)[1:2], collapse=','), nl-2, ifelse(nl>3, 's', ''))
  } else {
    x
  }
})

qc <- read_excel('~/Dropbox/CeMEE_v2_meta.xlsx', sheet = 'qc', range = "N1:O1000")
# QC column
maxhetfreq = 0.2
maxhetprop = 0.2
mincov = 0.1
addtag <- function(tag, boolv){
  o = matrix('PASS', nrow(meta))
  o[boolv] = tag
  o
}
qca = addtag('DUP', meta$line %in% qc$todrop.nobg[qc$case != 'haplotypeLL'])
qca <- cbind(qca, addtag('LOW', meta$`seq depth` < mincov))
qca <- cbind(qca, addtag('HAP', meta$line %in% qc$todrop.nobg[qc$case == 'haplotypeLL']))
qca <- cbind(qca, addtag('HET', meta$`perc ref/alt` > (maxhetfreq*100) & meta$`prop alt` > maxhetprop))
meta$`v2 QC` <- ''
meta$`v2 QC`[meta$v2==1] <- 'PASS'
dropix = apply(qca, 1, function(x) sum(x != 'PASS')>0)
meta$`v2 QC`[dropix] = unlist(apply(qca[dropix,], 1, function(x) paste0(unique(x[grep('PASS', x, invert = T)]), collapse=',')))
meta <- meta[,c(1:20,26,21:25)]

meta <- meta[,c(5,21,6:8,12,22:25)]

sketch = htmltools::withTags(table(
  class = "compact cell-border hover order-column stripe",
  thead(
    tr(
      th('Line', title = 'Unique identifier'),
      th('v2 QC', title = sprintf('Quality control status: PASS, DUP (another closely related line has been retained), LOW (sequencing coverage <%sx), HET (ref & alt seen at >%s%% of sites with >= 3 reads and mean minor allele proportion > %s)', mincov, maxhetfreq*100, maxhetprop)),
      th('Pop.', title = 'Population'),
      th('Pop. rep.', title = 'Population replicate'),
      th('Line number', title = 'Line number'),
      th('Seq depth', title = 'Sequenced depth (estimated from reads mapped to the center of chromosome I with minimum MQ40)'),
      th('Fert. NGM', title = 'Phenotyped for fertility (number of viable L1s) on NGM. Value is the number of observations (adult hermaphrodites) post-QC.'),
      th('Fert. NaCl', title = 'Fertility (number of viable L1s) on 300mM NaCl. Value is the number of observations (adult hermaphrodites) post-QC.'),
      th('Loco. NGM', title = 'Locomotion on NGM. Value is the number of observations (tracks) post-QC.'),
      th('Loco. NaCl', title = 'Locomotion on NaCl. Value is the number of observations (tracks) post-QC.')
    )
  )
))

datatable(meta, 
          container=sketch,
          rownames = F,
          options = list(pageLength = 25, 
                         autoWidth = T, 
                         scrollX=F,
                         initComplete = JS(
                           "function(settings, json) {",
                           "$(this.api().table().header()).css({'background-color': '#000', 'color': '#fff'});",
                           "}"
                         ),
                         columnDefs = list(
                           list(className = 'dt-center', targets='_all'),
                           list(width = '200px', targets='_all')
                         )
          ), 
          filter='top'
) %>% formatStyle('seq depth', background = styleColorBar(meta$`seq depth`, 'steelblue'))
```